### ISSUE TITLE
Examples on Apple M1 need an additional resolution in esy. 

### ISSUE THREAD

**Issue by pm-mck** [aantron/dream#160]
As of today, I found that running the examples in esy on an Apple M1 machine requires this resolution:

```json
  "resolutions": {
    "esy-openssl": "esy-packages/esy-openssl#619ae2d46ca981ec26ab3287487ad98b157a01d1"
  }
```

documented on the esy-openssl project page:

https://github.com/esy-packages/esy-openssl

If you don't do this, the esy build emits a few clang errors which will be difficult for a newcomer to handle.

Should this be added to the examples README or documented / provided in some other manner? Happy to open a pr but don't want to put it in the wrong place.

**Comment by aantron** [comment]
Can you make a PR that adds this to the resolutions of `example/2-middleware` here:

https://github.com/aantron/dream/blob/52d4da207ff2ae51ade431cc88ab26501a69cee7/example/2-middleware/esy.json#L11-L13

From your message, I suppose this will fix the example for you, and choosing this example for the PR will also get the change tested in CI, since that example is part of esy's quick start script that gets run in CI. If everything works, I'll do the busywork of adding this resolution to all the other examples.

@ManasJayanth is it correct to use the resolution this way? I assume that `esy-openssl#619ae2d46ca981ec26ab3287487ad98b157a01d1` will work on all platforms.

**Comment by sagarjs** [comment]
thank you for creating this @pm-mck
I was just about to create this myself though I have no idea how to fix the issue.

I tried adding the line to my esy.json in 2-middleware but I am getting the following error while running 'npx esy'
It looks like permission errors, but if i try manually creating the directory dynlinks as the logs seems to suggest, I do not get an error
what am I getting wrong here?

```
About the OCAML core installation:
        Standard library:      /Users/sagarsodah/.esy/3_____________________________________________________________/i/ocaml-4.12.0-44d44425/lib/ocaml
        Binaries:              /Users/sagarsodah/.esy/3_____________________________________________________________/i/ocaml-4.12.0-44d44425/bin
        Manual pages:          /Users/sagarsodah/.esy/3_____________________________________________________________/i/ocaml-4.12.0-44d44425/man
        Multi-threading type:  posix
    The directory of site-specific packages will be
        site-lib:              /Users/sagarsodah/.esy/3_____________________________________________________________/i/opam__s__ocaml_secondary_compiler-opam__c__4.08.1-1-eca62758/share/ocaml-secondary-compiler/lib
    The configuration file is written to:
        findlib config file:   /Users/sagarsodah/.esy/3_____________________________________________________________/i/ocaml-4.12.0-44d44425/etc/findlib.conf
    Software will be installed:
        Libraries:             in <site-lib>/findlib
        Binaries:              /Users/sagarsodah/.esy/3_____________________________________________________________/i/ocaml-4.12.0-44d44425/bin
        Manual pages:          /Users/sagarsodah/.esy/3_____________________________________________________________/i/ocaml-4.12.0-44d44425/man
        topfind script:        /Users/sagarsodah/.esy/3_____________________________________________________________/i/ocaml-4.12.0-44d44425/lib/ocaml
    Topfind ppxopt support:    yes
    Toolbox:                   no
    Link custom runtime:       yes
    Need bytes compatibility:  no
    
    Configuration has been written to Makefile.config
    
    You can now do 'make all', and optionally 'make opt', to build ocamlfind.
    # esy-build-package: running: 'make' 'install-meta'
    for x in dynlink stdlib unix bigarray compiler-libs ocamldoc threads str; do mkdir -p "/Users/sagarsodah/.esy/3_____________________________________________________________/i/opam__s__ocaml_secondary_compiler-opam__c__4.08.1-1-eca62758/share/ocaml-secondary-compiler/lib/$x"; cp site-lib-src/$x/META "/Users/sagarsodah/.esy/3_____________________________________________________________/i/opam__s__ocaml_secondary_compiler-opam__c__4.08.1-1-eca62758/share/ocaml-secondary-compiler/lib/$x"; done
    mkdir: /Users/sagarsodah/.esy/3_____________________________________________________________/i/opam__s__ocaml_secondary_compiler-opam__c__4.08.1-1-eca62758/share/ocaml-secondary-compiler/lib/dynlink: Operation not permitted
    cp: /Users/sagarsodah/.esy/3_____________________________________________________________/i/opam__s__ocaml_secondary_compiler-opam__c__4.08.1-1-eca62758/share/ocaml-secondary-compiler/lib/dynlink: Operation not permitted
    mkdir: /Users/sagarsodah/.esy/3_____________________________________________________________/i/opam__s__ocaml_secondary_compiler-opam__c__4.08.1-1-eca62758/share/ocaml-secondary-compiler/lib/stdlib: Operation not permitted
    cp: /Users/sagarsodah/.esy/3_____________________________________________________________/i/opam__s__ocaml_secondary_compiler-opam__c__4.08.1-1-eca62758/share/ocaml-secondary-compiler/lib/stdlib: Operation not permitted
    mkdir: /Users/sagarsodah/.esy/3_____________________________________________________________/i/opam__s__ocaml_secondary_compiler-opam__c__4.08.1-1-eca62758/share/ocaml-secondary-compiler/lib/unix: Operation not permitted
    cp: /Users/sagarsodah/.esy/3_____________________________________________________________/i/opam__s__ocaml_secondary_compiler-opam__c__4.08.1-1-eca62758/share/ocaml-secondary-compiler/lib/unix: Operation not permitted
    mkdir: /Users/sagarsodah/.esy/3_____________________________________________________________/i/opam__s__ocaml_secondary_compiler-opam__c__4.08.1-1-eca62758/share/ocaml-secondary-compiler/lib/bigarray: Operation not permitted
    cp: /Users/sagarsodah/.esy/3_____________________________________________________________/i/opam__s__ocaml_secondary_compiler-opam__c__4.08.1-1-eca62758/share/ocaml-secondary-compiler/lib/bigarray: Operation not permitted
    cp: /Users/sagarsodah/.esy/3_____________________________________________________________/i/opam__s__ocaml_secondary_compiler-opam__c__4.08.1-1-eca62758/share/ocaml-secondary-compiler/lib/compiler-libs/META: Operation not permitted
    cp: /Users/sagarsodah/.esy/3_____________________________________________________________/i/opam__s__ocaml_secondary_compiler-opam__c__4.08.1-1-eca62758/share/ocaml-secondary-compiler/lib/ocamldoc/META: Operation not permitted
    cp: /Users/sagarsodah/.esy/3_____________________________________________________________/i/opam__s__ocaml_secondary_compiler-opam__c__4.08.1-1-eca62758/share/ocaml-secondary-compiler/lib/threads/META: Operation not permitted
    mkdir: /Users/sagarsodah/.esy/3_____________________________________________________________/i/opam__s__ocaml_secondary_compiler-opam__c__4.08.1-1-eca62758/share/ocaml-secondary-compiler/lib/str: Operation not permitted
    cp: /Users/sagarsodah/.esy/3_____________________________________________________________/i/opam__s__ocaml_secondary_compiler-opam__c__4.08.1-1-eca62758/share/ocaml-secondary-compiler/lib/str: Operation not permitted
    make: *** [install-meta] Error 1
    error: command failed: 'make' 'install-meta' (exited with 2)
    esy-build-package: exiting with errors above...
    
  building @opam/ocamlfind-secondary@opam:1.9.1
esy: exiting due to errors above
```

**Comment by aantron** [comment]
@sagarjs This looks like an esy error (rather than Dream), and likely https://github.com/esy/esy/issues/1062. Can you post the full esy build log to https://github.com/esy/esy/issues/1062, including the part where it shows the `dune` and `ocaml` package versions built? cc @ManasJayanth

**Comment by aantron** [comment]
@sagarjs Please also include the esy version and the OS info, like at the [top of the esy issue](https://github.com/esy/esy/issues/1062#issue-566552098). Hopefully we can get this figured out quickly.

**Comment by sagarjs** [comment]
@aantron  sure, i am using ocaml@4.12.0 though; should I still post the error in that issue?

**Comment by ManasJayanth** [comment]
@aantron 

1. Yes, the resolution posted by @pm-mck is correct. And it just tested, esy-openssl builds fine on Windows, Macos M1 and Linux.

2. `ocaml-secondary` should never be depended on directly. Do you happen to recall why the examples suggest that users do so? Esy includes `ocaml-secondary` in the solution when the compiler is < 4.06 (There was a bug long ago where it was fetch for everyone, but has been fixed)

@sagarjs Can you try removing the `ocaml-secondary-compiler` and retry? With the following diff, the example built fine on my M1 mac

```diff
     "ocaml": "4.12.x"
   },
   "devDependencies": {
-    "@opam/ocaml-lsp-server": "*",
-    "@opam/ocamlfind-secondary": "*"
+    "@opam/ocaml-lsp-server": "*"
   },
   "resolutions": {
+    "esy-openssl": "esy-packages/esy-openssl#619ae2d46ca981ec26ab3287487ad98b157a01d1",
     "@opam/conf-libev": "esy-packages/libev:package.json#0b5eb6685b688649045aceac55dc559f6f21b829"
   },
 ```

**Comment by sagarjs** [comment]
@ManasJayanth  thanks!
This worked. Even though I did not add esy-openssl in the resolutions section

**Comment by ManasJayanth** [comment]
Without the override, openssl 1g released is pulled in on my M1 mac. So, I'd say the override is important. I wonder what `esy-openssl` is actually getting resolved to over there, @sagarjs . Would you be able to share the esy.lock/index.json?

**Comment by aantron** [comment]
@ManasJayanth Thanks!

`ocamlfind-secondary` was added to the resolutions in #65 by @dangdennis, with the [comment](https://github.com/aantron/dream/pull/65/files#diff-fa03cdae8368f8f9a6a73cfcc1226327735d610b35c887f2296bc511905403d7R8) that it is necessary for [OCaml Platform](https://github.com/ocamllabs/vscode-ocaml-platform) (the VSCode plugin) to do auto-completion. I don't know if that's accurate at this point &mdash; I am not currently using OCaml Platform, and haven't been able to set it up. @ManasJayanth, can you comment on the interaction between these resolutions and OCaml Platform?

**Comment by dangdennis** [comment]
With my recent try with Esy and the vscode plugin, I still had to use ocamlfind. Dune worked fine without it.

ah this is the error I encountered too. Use 1.8.1 ocamlfind/secondary!

Intel MacBook.

**Comment by dangdennis** [comment]
Removed ocamlfind and the vscode extension still works!

**Comment by sagarjs** [comment]
> Without the override, openssl 1g released is pulled in on my M1 mac (which doesn't have arm64 support). So, I'd say the override is important. I wonder what `esy-openssl` is actually getting resolved to over there, @sagarjs . Would you be able to share the esy.lock/index.json?

@ManasJayanth 
Pasting the esy-openssl part of the index.json below
I have installed opam using Rosetta2 Terminal because `brew install opam` did not work normally

```
    "esy-openssl@archive:https://github.com/openssl/openssl/archive/OpenSSL_1_1_1g.tar.gz#sha1:33324ff957edaae8ae575817b456320378da46ff@41b6fb3d": {
      "id":
        "esy-openssl@archive:https://github.com/openssl/openssl/archive/OpenSSL_1_1_1g.tar.gz#sha1:33324ff957edaae8ae575817b456320378da46ff@41b6fb3d",
      "name": "esy-openssl",
      "version":
        "archive:https://github.com/openssl/openssl/archive/OpenSSL_1_1_1g.tar.gz#sha1:33324ff957edaae8ae575817b456320378da46ff",
      "source": {
        "type": "install",
        "source": [
          "archive:https://github.com/openssl/openssl/archive/OpenSSL_1_1_1g.tar.gz#sha1:33324ff957edaae8ae575817b456320378da46ff"
        ]
      },
      "overrides": [ "esy.lock/overrides/41b6fb3db953254b0cb4c5e1b546f3f4" ],
      "dependencies": [
        "@opam/conf-pkg-config@opam:2@85f8644d",
        "@opam/conf-autoconf@github:esy-packages/esy-autoconf:package.json#fb93edf@d41d8cd9"
      ],
      "devDependencies": []
    },
```

**Comment by ManasJayanth** [comment]
I wonder if 1g works there because rosetta emulation - do you turn on intel emulation on while working on Dream (or in general)?

**Comment by sagarjs** [comment]
i have been running the examples without using the rosetta terminal
I have only run the examples till now

**Comment by sagarjs** [comment]
Also tried:
reinstalled opam using the binary distribution (without rosetta)
> installed esy dependencies again for example 2-middleware (without extra resolver and without ocamlfind-secondary)
> example runs properly

**Comment by sagarjs** [comment]
ok, did a clean re-install of opam (arch64)
|> clean ran npx esy in the example 2-middleware
|> got the error
|> added the resolution for esy-openssl
|> reran npx esy
|> worked perfectly

**Comment by aantron** [comment]
@sagarjs To confirm, you have it working with only adding the `esy-openssl` resolution, and you kept the `ocamlfind-secondary` dev dep, in your latest try?

**Comment by pm-mck** [comment]
I forgot to mention that I also had to remove `ocamlfind-secondary`. 

I just performed the following on my M1:

1. Remove `@opam/ocamlfind-secondary` from `2-middleware/esy.json`
2. Add `"esy-openssl": "esy-packages/esy-openssl#619...."` to `resolutions` in `2-middleware/esy.json`
3. Run `esy`
4. Run `esy start`
5. Wait for step 4 to print and then `curl http://localhost:8080`
6. Curl printed "Good morning, world!"

I think this was a successful test. I'll cut a PR in a few moments.
